src/dsa/rpo_falcon512/keys/secret_key.rs

```diff
@@ -0,0 +1,327 @@
+        let f_fft = f.fft();
+        let h_fft = g_fft.hadamard_div(&f_fft);
+        h_fft.ifft().into()
+    }
+
+    /// Signs a message polynomial with the secret key.
+    ///
+    /// Takes a randomness generator implementing `Rng` and message polynomial representing `c`
+    /// the hash-to-point of the message to be signed. It outputs a signature `Signature`.
+    ///
+    /// # Errors
+    /// Returns an error of signature generation fails.
+    fn sign_helper<R: Rng>(
+        &self,
+        c: Polynomial<FalconFelt>,
+        rng: &mut R,
+    ) -> Result<(SignaturePoly, SignaturePoly), FalconError> {
+        let one_over_q = 1.0 / (MODULUS as f64);
+        let c_over_q_fft = c.map(|cc| Complex::new(one_over_q * cc.value() as f64, 0.0)).fft();
+
+        // B = [[FFT(g), -FFT(f)], [FFT(G), -FFT(F)]]
+        let [g_fft, minus_f_fft, big_g_fft, minus_big_f_fft] = to_complex_fft(&self.secret_key);
+        let t0 = c_over_q_fft.hadamard_mul(&minus_big_f_fft);
+        let t1 = -c_over_q_fft.hadamard_mul(&minus_f_fft);
+
+        let s = loop {
+            let bold_s = loop {
+                let z = ffsampling(&(t0.clone(), t1.clone()), &self.tree, rng);
+                let t0_min_z0 = t0.clone() - z.0;
+                let t1_min_z1 = t1.clone() - z.1;
+
+                // s = (t-z) * B
+                let s0 = t0_min_z0.hadamard_mul(&g_fft) + t1_min_z1.hadamard_mul(&big_g_fft);
+                let s1 =
+                    t0_min_z0.hadamard_mul(&minus_f_fft) + t1_min_z1.hadamard_mul(&minus_big_f_fft);
+
+                // compute the norm of (s0||s1) and note that they are in FFT representation
+                let length_squared: f64 =
+                    (s0.coefficients.iter().map(|a| (a * a.conj()).re).sum::<f64>()
+                        + s1.coefficients.iter().map(|a| (a * a.conj()).re).sum::<f64>())
+                        / (N as f64);
+
+                if length_squared > (SIG_L2_BOUND as f64) {
+                    continue;
+                }
+
+                break [-s0, s1];
+            };
+            let s1 = bold_s[0].ifft();
+            let s2 = bold_s[1].ifft();
+            let s1_coef: [i16; N] = s1
+                .coefficients
+                .iter()
+                .map(|a| a.re.round() as i16)
+                .collect::<Vec<i16>>()
+                .try_into()
+                .expect("The number of coefficients should be equal to N");
+            let s2_coef: [i16; N] = s2
+                .coefficients
+                .iter()
+                .map(|a| a.re.round() as i16)
+                .collect::<Vec<i16>>()
+                .try_into()
+                .expect("The number of coefficients should be equal to N");
+
+            if let Ok(s1) = SignaturePoly::try_from(&s1_coef) {
+                if let Ok(s2) = SignaturePoly::try_from(&s2_coef) {
+                    if s2.fft().coefficients.iter().all(|&c| c != FalconFelt::zero()) {
+                        break (s1, s2);
```

Comment:
I think this can probably be just `return Ok((s1, s2));`.

Reply:
Changed